### ISSUE TITLE
Using `Strict` constructor does not appear to always force the contents to WHNF, leading to segfaults.

### ISSUE THREAD

**Issue by dcoutts** [sgraf812/data-elevator#4]
With the latest version of `data-elevator` (0.1.0.2) I run into segfaults. I was not getting them before with the same code with an earlier version of `data-elevator` (0.1.0.1). I know very little changed between those versions so that's surprising. I'm using ghc-9.6.3 (Yes I had to ignore the upper bound on base to make version 0.1.0.1 work.

Here's the code that does _not_ work:
```
primArrayToStrictArray# :: forall a s.
                           P.Array a
                        -> GHC.State# s
                        -> (# GHC.State# s, GHC.Array# (Strict a) #)
primArrayToStrictArray# (P.Array a) s0 =
    case GHC.newArray# (GHC.sizeofArray# a) (Strict (P.indexArray (P.Array a) 0)) s0 of
      (# s1, a#  #) ->
        case go a# 0# s1 of
          s2 -> GHC.unsafeFreezeArray# a# s2
  where
    go :: GHC.MutableArray# s (Strict a)
       -> GHC.Int#
       -> GHC.State# s
       -> GHC.State# s
    go a# i# s =
      case i# GHC.<# GHC.sizeofArray# a of
        1# -> case GHC.indexArray# a i# of
                (# x #) ->
                  x `seq` case GHC.writeArray# a# i# (Strict x) s of
                            s' -> go a# (i# +# 1#) s'
        _  -> s
```
Note in particular this bit:
```
case GHC.indexArray# a i# of
  (# x #) ->
    x `seq` case GHC.writeArray# a# i# (Strict x) s of
      s' -> go a# (i# +# 1#) s'
```
I'm using `seq` here in a way that should be superfluous, because `Strict` itself should be seen by ghc as being strict. Looking t the core, I don't see any `case` analysis on the `x` being written into the array, so I think it is not in fact being evaluated to WHNF.
And indexing the resulting array results in segfaults. I presume what happens is that the thunk is interpreted as a constructor which then contains junk. I see this with the constructors I'm using, the non-pointer fields contain junk values and the pointer fields provoke segfaults.

I can "fix" this by hitting it with the "no really really evaluate it please" primop `seq#` which doesn't get eliminated by the optimiser.
```
primArrayToStrictArray# :: forall a s.
                           P.Array a
                        -> GHC.State# s
                        -> (# GHC.State# s, GHC.Array# (Strict a) #)
primArrayToStrictArray# (P.Array a) s0 =
    case GHC.newArray# (GHC.sizeofArray# a) (Strict (P.indexArray (P.Array a) 0)) s0 of
      (# s1, a#  #) ->
        case go a# 0# s1 of
          s2 -> GHC.unsafeFreezeArray# a# s2
  where
    go :: GHC.MutableArray# s (Strict a)
       -> GHC.Int#
       -> GHC.State# s
       -> GHC.State# s
    go a# i# s =
      case i# GHC.<# GHC.sizeofArray# a of
        1# -> case GHC.indexArray# a i# of
                (# x #) ->
                  case GHC.seq# x s of
                    (# s', x' #) -> case GHC.writeArray# a# i# (Strict x') s' of
                              s'' -> go a# (i# +# 1#) s''
        _  -> s
```
This then works. No segfaults and the resulting array contains the contents I expect.

**Comment by sgraf812** [comment]
Thanks! This is quite interesting and perhaps confirms my fear that this library is unsound in some cases. I have a few questions, though:

1. Did your code work with 0.1.0.1 *without* the `seq`? (In a perfect levity-polymorphic world this `seq` would be redundant altogether because it *must* be inserted by the compiler.)
2. Can you post the optimised Core, in particular including demand signatures?

The general problem with this library is that GHC is free to float around the coercion that represents the `Strict` constructor. But since this library wants `Strict` to translate into a `seq` in CorePrep for the atomic `x` in your example, the coercion is no longer zero-cost and thus violates one of the basic principles of Coercions as implemented in GHC.

In your case the reason might be different, though, and represent an actual bug in GHC. Until I see Core it's all just guesses, though.

**Comment by sgraf812** [comment]
Another workaround could be to use a non-zero cost conversion function

```hs
reallyStrict :: a -> Strict a
reallyStrict !a = Strict a
{-# NOINLINE reallyStrict #-}
```

It appears that this version is necessary whenever the argument to `Strict` is atomic (i.e. a variable). Much like `seq#`, this function leads to a function call and an eval that cannot cancel away. Furthermore, unlike `Strict x`, `reallyStrict x` is not atomic itself, so it *will* be evaluated in CorePrep prior to the call to `writeArray#`.

Perhaps the bang in `reallyStrict` is not even needed (feel free to try) and all we need is a way to tell GHC that "`Strict x` is not atomic and thus must be evaluted before the call to `writeArray#`".

**Comment by dcoutts** [comment]
Yes it did work before without the seq. I added that just for emphasis / testing.

I'll post the core later.

Yes I imagine the `reallyStrict` would do much the same as the `seq#`, in that it would not get eliminated by the optimiser.

**Comment by dcoutts** [comment]
Ok, here's the basic `-ddump-simpl`.
```
-- RHS size: {terms: 65, types: 152, coercions: 16, joins: 0/1}
Database.LSMTree.Internal.StrictArray.$wprimArrayToStrictArray [InlPrag=[2]]
  :: forall {a}. GHC.Array# a -> GHC.Array# (Strict a)
[GblId,
 Arity=1,
 Str=<L>,
 Unf=Unf{Src=<vanilla>, TopLvl=True,
         Value=True, ConLike=True, WorkFree=True, Expandable=True,
         Guidance=IF_ARGS [0] 183 0}]
Database.LSMTree.Internal.StrictArray.$wprimArrayToStrictArray
  = \ (@a_s4qY) (ww_s4r1 :: GHC.Array# a_s4qY) ->
      GHC.runRW#
        @GHC.UnliftedRep
        @(GHC.Array# (Strict a_s4qY))
        (\ (s_i4q2 [OS=OneShot] :: GHC.State# GHC.RealWorld) ->
           case GHC.newArray#
                  @GHC.Unlifted
                  @(Strict a_s4qY)
                  @GHC.RealWorld
                  (GHC.sizeofArray# @GHC.Lifted @a_s4qY ww_s4r1)
                  (case Unsafe.Coerce.unsafeEqualityProof
                          @(*) @(GHC.Any -> GHC.Any) @(a_s4qY -> Strict a_s4qY)
                   of
                   { Unsafe.Coerce.UnsafeRefl co_a4pL ->
                   case GHC.indexArray# @GHC.Lifted @a_s4qY ww_s4r1 0# of
                   { (# ipv_a4qh #) ->
                   ipv_a4qh
                   `cast` (SelCo:Fun(arg) (Sub co_a4pL)
                           ; SelCo:Fun(res) (Sub (Sym co_a4pL))
                           :: a_s4qY ~R# Strict a_s4qY)
                   }
                   })
                  s_i4q2
           of
           { (# ipv_s4qj, ipv1_s4qk #) ->
           letrec {
             go_s4qQ [Occ=LoopBreaker, Dmd=SC(S,C(1,C(1,L)))]
               :: forall s1.
                  GHC.MutableArray# s1 (Strict a_s4qY)
                  -> GHC.Int# -> GHC.State# s1 -> GHC.State# s1
             [LclId, Arity=3, Str=<L><L><L>, Unf=OtherCon []]
             go_s4qQ
               = \ (@s1_a4ne)
                   (a#_a4kk :: GHC.MutableArray# s1_a4ne (Strict a_s4qY))
                   (i#_a4kl :: GHC.Int#)
                   (s2_a4km :: GHC.State# s1_a4ne) ->
                   case GHC.<# i#_a4kl (GHC.sizeofArray# @GHC.Lifted @a_s4qY ww_s4r1)
                   of {
                     __DEFAULT -> s2_a4km;
                     1# ->
                       case GHC.indexArray# @GHC.Lifted @a_s4qY ww_s4r1 i#_a4kl of
                       { (# ipv2_s4pz #) ->
                       case GHC.writeArray#
                              @GHC.Unlifted
                              @s1_a4ne
                              @(Strict a_s4qY)
                              a#_a4kk
                              i#_a4kl
                              (case Unsafe.Coerce.unsafeEqualityProof
                                      @(*) @(GHC.Any -> GHC.Any) @(a_s4qY -> Strict a_s4qY)
                               of
                               { Unsafe.Coerce.UnsafeRefl co_a4pL ->
                               ipv2_s4pz
                               `cast` (SelCo:Fun(arg) (Sub co_a4pL)
                                       ; SelCo:Fun(res) (Sub (Sym co_a4pL))
                                       :: a_s4qY ~R# Strict a_s4qY)
                               })
                              s2_a4km
                       of s'_a4ko
                       { __DEFAULT ->
                       go_s4qQ @s1_a4ne a#_a4kk (+# i#_a4kl 1#) s'_a4ko
                       }
                       }
                   }; } in
           case go_s4qQ @GHC.RealWorld ipv1_s4qk 0# ipv_s4qj of s2_a4ks
           { __DEFAULT ->
           case GHC.unsafeFreezeArray#
                  @GHC.Unlifted @GHC.RealWorld @(Strict a_s4qY) ipv1_s4qk s2_a4ks
           of
           { (# ipv2_s4qm, ipv3_s4qn #) ->
           ipv3_s4qn
           }
           }
           })
```

**Comment by dcoutts** [comment]
And `-dump-stranal` with just the core loop `go`:
```
 go_s4qQ
   = \ (@s_a4ne)
       (a#_a4kk :: GHC.MutableArray# s_a4ne (Strict a_a4mO))
       (i#_a4kl :: GHC.Int#)
       (s_a4km :: GHC.State# s_a4ne) ->
       case GHC.<# i#_a4kl (GHC.sizeofArray# @GHC.Lifted @a_a4mO a_a4kh)
       of {
         __DEFAULT -> s_a4km;
         1# ->
           case GHC.indexArray# @GHC.Lifted @a_a4mO a_a4kh i#_a4kl of
           { (# ipv_s4pz [Dmd=1L] #) ->
           case GHC.writeArray#
                  @GHC.Unlifted
                  @s_a4ne
                  @(Strict a_a4mO)
                  a#_a4kk
                  i#_a4kl
                  (case Unsafe.Coerce.unsafeEqualityProof
                          @(*) @(GHC.Any -> GHC.Any) @(a_a4mO -> Strict a_a4mO)
                   of
                   { Unsafe.Coerce.UnsafeRefl co_a4pL ->
                   ipv_s4pz
                   `cast` (SelCo:Fun(arg) (Sub co_a4pL)
                           ; SelCo:Fun(res) (Sub (Sym co_a4pL))
                           :: a_a4mO ~R# Strict a_a4mO)
                   })
                  s_a4km
           of s'_a4ko
           { __DEFAULT ->
           go_s4qQ @s_a4ne a#_a4kk (+# i#_a4kl 1#) s'_a4ko
           }
           }
       };
```

Note that both of these versions are without any use of `seq` or `seq#`. The source is:
```
    go :: forall s.
          GHC.MutableArray# s (Strict a)
       -> GHC.Int#
       -> GHC.State# s
       -> GHC.State# s
    go a# i# s =
      case i# GHC.<# GHC.sizeofArray# a of
        1# ->
          case GHC.indexArray# a i# of
            (# x #) ->
              -- We have to use seq# here to force the array element to WHNF
              -- before putting it into the strict array. This should not be
              -- necessary. https://github.com/sgraf812/data-elevator/issues/4
{-              case GHC.seq# x s of
                (# s', x' #) ->-}
                  case GHC.writeArray# a# i# (Strict x)  s of
                    s' -> go a# (i# +# 1#) s'
        _  -> s
```
(see `seq#` commented out and s'' vs s' fixed up correspondingly)

**Comment by dcoutts** [comment]
Standalone (hopefully) code for you to try:

https://gist.github.com/dcoutts/e2eda75b9961d2a27149bbf86edcedf6

**Comment by sgraf812** [comment]
Thanks. I think this confirms my hunch. ~~I will try to find a separate reproducer~~ (thanks for providing one), but the general problem is that

```
case Unsafe.Coerce.unsafeEqualityProof
                          @(*) @(GHC.Any -> GHC.Any) @(a_a4mO -> Strict a_a4mO)
                   of
                   { Unsafe.Coerce.UnsafeRefl co_a4pL ->
                   ipv_s4pz
                   `cast` (SelCo:Fun(arg) (Sub co_a4pL)
                           ; SelCo:Fun(res) (Sub (Sym co_a4pL))
                           :: a_a4mO ~R# Strict a_a4mO)
                   }
```

is discovered by GHC as trivial, in which case it does not case-bind it during CorePrep (I should have asked for  `-ddump-stg-final` instead, just to be sure ...).

Indeed I think a safer, but non-zero-cost design of the library would be to make the PatSyn `Strict` `NOINLINE` and instead declare rewrite RULEs `Strict . Lazy = id`, `Lazy . Strict = id` (presumably that also needs `Lazy` to be `NOINLINE`, but let's see).

It has become clear that a safe zero-cost design needs compiler support. For example, by inlining `Strict` during CorePrep, akin to `seq#` (which nowadays has a definition in `GHC.IO.Internal`).

**Comment by sgraf812** [comment]
I fixed this issue on `main`, added repro test and released [0.2](https://hackage.haskell.org/package/data-elevator-0.2) with it.

I now also know why this wasn't a problem before 9.6: It has to do with abolishing the let/app invariant. Specifically, in 9.4 and below, we desugared `f (Strict x)` to `case Strict x of y { __DEFAULT -> f y }`, with 9.6 we allow `f (Strict x)` as well. In the former, we keep around the Case until after CorePrep, which is what we want, whereas in the latter it ultimately inlines into the trivial `f (x |> co)` for which CorePrep does not introduce an eval before the call (perhaps its should).

The fix of making `toStrict#` `NOINLINE` works fine in 9.6 and above where I can also introduce rewrite rules such as `fromStrict# (toStrict# x) = x` to counter perf implications and uphold the near zero-cost promise. However, this rewrite rule is only expressible *because* we abolished the let/app invariant in 9.6! So we cannot have this rewrite RULE in 9.4 and below, so 0.2 can be slower there than 0.1.

Now the question for you, Duncan, is: **Is absolute zero-cost perf necessary for your client in 9.4 and below?** Then you might want to continue using the `0.1.*` series there. Otherwise, and if you plan to support 9.6 exclusively anway, just use the freshly released 0.2.
